src/main/java/com/google/api/codegen/transformer/ApiMethodTransformer.java

```diff
@@ -0,0 +1,260 @@
+
+    docBuilder.mainDocLines(namer.getDocLines(context.getMethod()));
+    List<ParamDocView> paramDocs =
+        getMethodParamDocs(context, context.getMethodConfig().getRequiredFields());
+    paramDocs.add(getOptionalArrayParamDoc(context, context.getMethodConfig().getOptionalFields()));
+    paramDocs.add(getCallSettingsParamDoc(context));
+    docBuilder.paramDocs(paramDocs);
+    docBuilder.returnTypeName(
+        namer.getDynamicReturnTypeName(
+            context.getTypeTable(), context.getMethod(), context.getMethodConfig()));
+    docBuilder.throwsDocLines(new ArrayList<String>());
+
+    apiMethod.doc(docBuilder.build());
+
+    apiMethod.name(namer.getApiMethodName(context.getMethod()));
+    apiMethod.requestTypeName(
+        context.getTypeTable().getAndSaveNicknameFor(context.getMethod().getInputType()));
+    apiMethod.hasReturnValue(!ServiceMessages.s_isEmptyType(context.getMethod().getOutputType()));
+    apiMethod.key(namer.getMethodKey(context.getMethod()));
+    apiMethod.grpcMethodName(namer.getGrpcMethodName(context.getMethod()));
+
+    List<DynamicDefaultableParamView> methodParams = new ArrayList<>();
+    for (Field field : context.getMethodConfig().getRequiredFields()) {
+      methodParams.add(generateDefaultableParam(context, field));
+    }
+
+    // TODO create a map TypeRef here instead of an array
+    // (not done yet because array is sufficient for PHP, and maps are more complex to construct)
+    TypeRef arrayType = TypeRef.fromPrimitiveName("string").makeRepeated();
+
+    DynamicDefaultableParamView.Builder optionalArgs = DynamicDefaultableParamView.newBuilder();
+    optionalArgs.name(namer.varName(Name.from("optional", "args")));
+    optionalArgs.defaultValue(context.getTypeTable().getZeroValueAndSaveNicknameFor(arrayType));
+    methodParams.add(optionalArgs.build());
+
+    DynamicDefaultableParamView.Builder callSettings = DynamicDefaultableParamView.newBuilder();
+    callSettings.name(namer.varName(Name.from("call", "settings")));
+    callSettings.defaultValue(context.getTypeTable().getZeroValueAndSaveNicknameFor(arrayType));
+    methodParams.add(callSettings.build());
+
+    apiMethod.methodParams(methodParams);
+
+    List<RequestObjectParamView> requiredRequestObjectParams = new ArrayList<>();
+    for (Field field : context.getMethodConfig().getRequiredFields()) {
+      requiredRequestObjectParams.add(generateRequestObjectParam(context, field));
+    }
+    apiMethod.requiredRequestObjectParams(requiredRequestObjectParams);
+
+    List<RequestObjectParamView> optionalRequestObjectParams = new ArrayList<>();
+    for (Field field : context.getMethodConfig().getOptionalFields()) {
+      optionalRequestObjectParams.add(generateRequestObjectParam(context, field));
+    }
+    apiMethod.optionalRequestObjectParams(optionalRequestObjectParams);
+
+    return apiMethod.build();
+  }
+
+  public DynamicDefaultableParamView generateDefaultableParam(
```

Comment:
Do these methods (generateDefaultableParam and generateRequestObjectParam) need to be public? And, could they accept an Iterable<Field> and return a List, like getMethodParamDocs? This is only a nit, but makes the generateOptionalArrayMethod closer to being all one-liner apiMethod.<something>() calls.